File: Classes/Document-based code/PPDocument.m
//
//  PPDocument.m
//  PPMacho
//
//  Created by C.W. Betts on 9/7/13.
//
//

#import "PPDocument.h"
#import <PlayerPROKit/PlayerPROKit.h>
#import <AVFoundation/AVFoundation.h>
#import "PPApp_AppDelegate.h"
#import "UserDefaultKeys.h"
#import "PPExportObject.h"
#import "PPErrors.h"
#import "BoxViewController.h"

@interface PPDocument ()
@property (strong) PPDriver *theDriver;
@property (strong) PPMusicObjectWrapper *theMusic;
@property MADDriverSettings exportSettings;
@end

@implementation PPDocument
@synthesize exportSettings;

- (NSString*)musicName
{
	return self.theMusic.internalFileName;
}

- (void)setMusicName:(NSString *)musicName
{
	[self willChangeValueForKey:@"musicName"];
	self.theMusic.internalFileName = musicName;
	[self didChangeValueForKey:@"musicName"];
}

- (NSString*)musicInfo
{
	return self.theMusic.madInfo;
}

- (void)setMusicInfo:(NSString *)musicInfo
{
	[self willChangeValueForKey:@"musicInfo"];
	self.theMusic.madInfo = musicInfo;
	[self didChangeValueForKey:@"musicInfo"];
}

- (NSString*)authorString
{
	return self.theMusic.madAuthor;
}

- (void)setAuthorString:(NSString *)authorString
{
	[self willChangeValueForKey:@"authorString"];
	self.theMusic.madAuthor = authorString;
	[self didChangeValueForKey:@"authorString"];
}

- (void)soundPreferencesDidChange:(NSNotification *)notification
{
	[self MADDriverWithPreferences];
}

- (id)init
{
	self = [super init];
	if (self) {
		MADDriverSettings init;
		{
			MADGetBestDriver(&init);
			NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
			
			//TODO: Sanity Checking
			init.surround = [defaults boolForKey:PPSurroundToggle];
			init.outPutRate = (unsigned int)[defaults integerForKey:PPSoundOutRate];
			init.outPutBits = [defaults integerForKey:PPSoundOutBits];
			if ([defaults boolForKey:PPOversamplingToggle]) {
				init.oversampling = (int)[defaults integerForKey:PPOversamplingAmount];
			} else {
				init.oversampling = 1;
			}
			init.Reverb = [defaults boolForKey:PPReverbToggle];
			init.ReverbSize = (int)[defaults integerForKey:PPReverbAmount];
			init.ReverbStrength = (int)[defaults integerForKey:PPReverbStrength];
			if ([defaults boolForKey:PPStereoDelayToggle]) {
				init.MicroDelaySize = (int)[defaults integerForKey:PPStereoDelayAmount];
			} else {
				init.MicroDelaySize = 0;
			}
			
			init.driverMode = [defaults integerForKey:PPSoundDriver];
			init.repeatMusic = FALSE;
		}
		
		self.theDriver = [[PPDriver alloc] initWithLibrary:globalMadLib settings:&init];
		
		self.exportController = [[PPSoundSettingsViewController alloc] init];
		self.exportController.delegate = self;
		
		NSNotificationCenter *defaultCenter = [NSNotificationCenter defaultCenter];
		[defaultCenter addObserver:self selector:@selector(soundPreferencesDidChange:) name:PPSoundPreferencesDidChange object:nil];
		
	}
	return self;
}

- (NSString *)windowNibName
{
	// Override returning the nib file name of the document
	// If you need to use a subclass of NSWindowController or if your document supports multiple NSWindowControllers, you should remove this method and override -makeWindowControllers instead.
	return @"PPDocument";
}

- (void)windowControllerDidLoadNib:(NSWindowController *)aController
{
	[super windowControllerDidLoadNib:aController];
	// Add any code here that needs to be executed once the windowController has loaded the document's window.
}

- (NSFileWrapper *)fileWrapperOfType:(NSString *)typeName error:(NSError **)outError
{
	if (outError) {
		*outError = nil;
	}
	return self.theMusic.musicWrapper;
}

- (BOOL)readFromFileWrapper:(NSFileWrapper *)fileWrapper ofType:(NSString *)typeName error:(NSError **)outError
{
	self.theMusic = [[PPMusicObjectWrapper alloc] initWithFileWrapper:fileWrapper];
	if (self.theMusic) {
		if (outError) {
			*outError = nil;
		}
		
		return YES;
	} else {
		if (outError) {
			*outError = CreateErrorFromMADErrorType(MADReadingErr);
		}
		
		return NO;
	}
}

+ (BOOL)autosavesInPlace
{
	return YES;
}

- (void)MADDriverWithPreferences
{
	OSErr returnerr = noErr;
	MADDriverSettings init;
	MADGetBestDriver(&init);
	NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
	
	//TODO: Sanity Checking
	init.surround = [defaults boolForKey:PPSurroundToggle];
	init.outPutRate = (unsigned int)[defaults integerForKey:PPSoundOutRate];
	init.outPutBits = [defaults integerForKey:PPSoundOutBits];
	if ([defaults boolForKey:PPOversamplingToggle]) {
		init.oversampling = (int)[defaults integerForKey:PPOversamplingAmount];
	} else {
		init.oversampling = 1;
	}
	init.Reverb = [defaults boolForKey:PPReverbToggle];
	init.ReverbSize = (int)[defaults integerForKey:PPReverbAmount];
	init.ReverbStrength = (int)[defaults integerForKey:PPReverbStrength];
	if ([defaults boolForKey:PPStereoDelayToggle]) {
		init.MicroDelaySize = (int)[defaults integerForKey:PPStereoDelayAmount];
	} else {
		init.MicroDelaySize = 0;
	}
	
	init.driverMode = [defaults integerForKey:PPSoundDriver];
	init.repeatMusic = FALSE;
	
	returnerr = [_theDriver changeDriverSettingsToSettings:init];
	//[[NSNotificationCenter defaultCenter] postNotificationName:PPDriverDidChange object:self];
	
	if (returnerr != noErr) {
		[[NSAlert alertWithError:CreateErrorFromMADErrorType(returnerr)] beginSheetModalForWindow:[self windowForSheet] completionHandler:^(NSModalResponse returnCode) {
			;//Currently, do nothing
		}];
		//return;
	}
}

- (void)importMusicObject:(PPMusicObject*)theObj
{
	[self importMusicObjectWrapper:[[PPMusicObjectWrapper alloc] initFromMusicObject:theObj]];
}

- (void)importMusicObjectWrapper:(PPMusicObjectWrapper*)theWrap
{
	if (!_theMusic) {
		self.theMusic = theWrap;
	}
}

#if 0
- (void)updateMusicStats:(NSTimer*)theTimer
{
	if (music) {
		long fT, cT;
		MADGetMusicStatus(madDriver, &fT, &cT);
		if (MADIsDonePlaying(madDriver) && !self.paused && !MADIsExporting(madDriver)) {
			[self songIsDonePlaying];
			MADGetMusicStatus(madDriver, &fT, &cT);
		}
		[songPos setDoubleValue:cT];
		[songCurTime setIntegerValue:cT];
	}
}

- (IBAction)saveInstrumentList:(id)sender
{
	[_theDriver beginExport];
	NSSavePanel *savePanel = [NSSavePanel savePanel];
	[savePanel setAllowedFileTypes:@[PPInstrumentListUTI]];
	[savePanel setCanCreateDirectories:YES];
	[savePanel setCanSelectHiddenExtension:YES];
	if (![musicName isEqualToString:@""]) {
		[savePanel setNameFieldStringValue:[NSString stringWithFormat:@"%@'s instruments", musicName]];
	} else {
		[savePanel setNameFieldStringValue:@"Tracker Instruments"];
	}
	
	if ([savePanel runModal] == NSFileHandlingPanelOKButton) {
		OSErr fileErr = [instrumentController exportInstrumentListToURL:[savePanel URL]];
		if (fileErr) {
			NSError *theErr = CreateErrorFromMADErrorType(fileErr);
			[[NSAlert alertWithError:theErr] runModal];
		}
	}
	
	[_theDriver endExport];
}

#endif

static inline extended80 convertSampleRateToExtended80(unsigned int theNum)
{
	//Yes, the pragma pack is needed
	//otherwise the data will be improperly mapped.
#pragma pack(push, 2)
	union {
		extended80 shortman;
		struct Float80i {
			SInt16  exp;
			UInt32  man[2];
		} intman;
	} toreturn;
#pragma pack(pop)
	
	unsigned int shift, exponent;
	
	for(shift = 0U; (theNum >> (31 - shift)) == 0U; ++shift)
		;
	theNum <<= shift;
	exponent = 63U - (shift + 32U); /* add 32 for unused second word */
	
	toreturn.intman.exp = (exponent+0x3FFF);
	PPBE16(&toreturn.intman.exp);
	toreturn.intman.man[0] = theNum;
	PPBE32(&toreturn.intman.man[0]);
	toreturn.intman.man[1] = 0;
	PPBE32(&toreturn.intman.man[1]);
	
	return toreturn.shortman;
}

- (NSMutableData *)rawSoundData:(MADDriverSettings*)theSet
{
	OSErr err = noErr;
	PPDriver *theRec = [[PPDriver alloc] initWithLibrary:globalMadLib settings:theSet error:&err];
	
	if (theRec == nil) {
		dispatch_async(dispatch_get_main_queue(), ^{
			NSError *NSerr = CreateErrorFromMADErrorType(err);
			[[NSAlert alertWithError:NSerr] beginSheetModalForWindow:[self windowForSheet] completionHandler:^(NSModalResponse returnCode) {
				;//Do nothing for now
			}];
		});
		
		return nil;
	}
	
	[theRec cleanDriver];
	[_theMusic attachToDriver:theRec];
	
	char *soundPtr = NULL;
	NSUInteger full = [theRec audioLength];
	if (theSet->outPutBits == 16) {
		full *= 2;
	} else if (theSet->outPutBits == 20 || theSet->outPutBits == 24 ) {
		full *= 3;
	}
	
	switch (theSet->outPutMode) {
		case DeluxeStereoOutPut:
		case StereoOutPut:
			full *= 2;
			break;
			
		case PolyPhonic:
			full *= 4;
			break;
			
		default:
			break;
	}
	
	[theRec play];
	
	NSMutableData *mutData = [[NSMutableData alloc] init];
	soundPtr = calloc(full, 1);
	
	while([theRec directSaveToPointer:soundPtr settings:theSet])
		[mutData appendBytes:soundPtr length:full];
	
	free(soundPtr);
	return mutData;
}

#import "getAIFF.i"

typedef struct {
	NSInteger	tag;
	CFURLRef	fileURL;
} ExportContextInfo;

- (void)exportSheetDidEnd:(NSWindow *)sheet returnCode:(NSInteger)returnCode contextInfo:(void *)contextInfo;
{
	ExportContextInfo *ei = contextInfo;
	NSURL *outURL = CFBridgingRelease(ei->fileURL);
	
	if (returnCode == NSAlertDefaultReturn) {
		switch (ei->tag)
		{
			case -1:
				//AIFF
			{
				PPExportObject *expObj = [[PPExportObject alloc] initWithDestination:outURL exportBlock:^OSErr(NSURL *theURL, NSString * __autoreleasing *errStr) {
					if (errStr)
						*errStr = nil;

					OSErr theErr = [self saveMusicAsAIFFToURL:theURL usingSettings:&exportSettings];
					[_theDriver endExport];
					return theErr;
				}];
				[[NSApp delegate] addExportObject:expObj];
			}
				break;
				
			case -2:
				//M4A
			{
				PPExportObject *expObj = [[PPExportObject alloc] initWithDestination:outURL exportBlock:^OSErr(NSURL *theURL, NSString * __autoreleasing *errStr) {
					OSErr theErr = noErr;
					if (errStr)
						*errStr = nil;
					
					NSArray *metadataInfo;
					NSString *oldMusicName = self.musicName;
					NSString *oldMusicInfo = self.musicInfo;
					NSURL *oldURL = [self fileURL];
					NSURL *tmpURL = [[[NSFileManager defaultManager] URLForDirectory:NSItemReplacementDirectory inDomain:NSUserDomainMask appropriateForURL:oldURL create:YES error:NULL] URLByAppendingPathComponent:[NSString stringWithFormat:@"%@.aiff", (oldMusicName && ![oldMusicName isEqualToString:@""]) ? oldMusicName : @"untitled"] isDirectory:NO];
					theErr = [self saveMusicAsAIFFToURL:tmpURL usingSettings:&exportSettings];
					[_theDriver endExport];
					if (theErr) {
						if (errStr)
							*errStr = [[NSString alloc] initWithFormat:@"Unable to save temporary file to %@, error %d.", [tmpURL path], theErr];

						return theErr;
					}
					
					AVAsset *exportMov = [AVAsset assetWithURL:tmpURL];
					
					{
						AVMutableMetadataItem *titleName = [AVMutableMetadataItem new];
						[titleName setKeySpace:AVMetadataKeySpaceCommon];
						[titleName setKey:AVMetadataCommonKeyTitle];
						[titleName setValue:oldMusicName];
						
						AVMutableMetadataItem *dataInfo = [AVMutableMetadataItem new];
						[titleName setKeySpace:AVMetadataKeySpaceCommon];
						[titleName setKey:AVMetadataCommonKeySoftware];
						[titleName setValue:@"PlayerPRO 6"];
						
						AVMutableMetadataItem *musicInfoQTUser = [AVMutableMetadataItem new];
						[musicInfoQTUser setKeySpace:AVMetadataKeySpaceQuickTimeUserData];
						[musicInfoQTUser setKey:AVMetadataQuickTimeUserDataKeyInformation];
						[musicInfoQTUser setValue:oldMusicInfo];
						
						AVMutableMetadataItem *musicInfoiTunes = [AVMutableMetadataItem new];
						[musicInfoiTunes setKeySpace:AVMetadataKeySpaceiTunes];
						[musicInfoiTunes setKey:AVMetadataiTunesMetadataKeyUserComment];
						[musicInfoiTunes setValue:oldMusicInfo];
						
						AVMutableMetadataItem *musicInfoQTMeta = [AVMutableMetadataItem new];
						[musicInfoQTMeta setKeySpace:AVMetadataKeySpaceQuickTimeMetadata];
						[musicInfoQTMeta setKey:AVMetadataQuickTimeMetadataKeyInformation];
						[musicInfoQTMeta setValue:oldMusicInfo];
						
						
						metadataInfo = @[titleName, dataInfo, musicInfoQTUser, musicInfoiTunes, musicInfoQTMeta];
					}
					
					oldMusicInfo = nil;
					if (!exportMov) {
						if (errStr) {
							*errStr = [[NSString alloc] initWithFormat:@"Init Failed for %@.", oldMusicName];
						}
						[[NSFileManager defaultManager] removeItemAtURL:tmpURL error:NULL];
						return MADWritingErr;
					}
					
					AVAssetExportSession *session = [[AVAssetExportSession alloc] initWithAsset:exportMov presetName:AVAssetExportPresetAppleM4A];
					if (!session) {
						if (errStr)
							*errStr = [[NSString alloc] initWithFormat:@"Export session creation for %@ failed.", oldMusicName];
						
						[[NSFileManager defaultManager] removeItemAtURL:tmpURL error:NULL];
						return MADWritingErr;
					}
					[[NSFileManager defaultManager] removeItemAtURL:theURL error:NULL];
					session.outputURL = theURL;
					session.outputFileType = AVFileTypeAppleM4A;
					session.metadata = metadataInfo;
					metadataInfo = nil;
					dispatch_semaphore_t sessionWaitSemaphore = dispatch_semaphore_create(0);
					[session exportAsynchronouslyWithCompletionHandler:^{
						dispatch_semaphore_signal(sessionWaitSemaphore);
					}];
					dispatch_semaphore_wait(sessionWaitSemaphore, DISPATCH_TIME_FOREVER);
					
					BOOL didFinish = [session status] == AVAssetExportSessionStatusCompleted;
					[[NSFileManager defaultManager] removeItemAtURL:tmpURL error:NULL];
					
					if (didFinish) {
						return noErr;
					} else {
						if (errStr)
							*errStr = [[NSString alloc] initWithFormat:@"export of \"%@\" failed.", oldMusicName];
						
						return MADWritingErr;
					}
				}];
				[[NSApp delegate] addExportObject:expObj];
			}
				break;
				
			default:
				break;
		}
	}
	[_theDriver endExport];
	
	free(contextInfo);
}

- (void)showExportSettingsWithExportType:(NSInteger)expType URL:(NSURL*)theURL
{
	MADGetBestDriver(&exportSettings);
	exportSettings.driverMode = NoHardwareDriver;
	exportSettings.repeatMusic = FALSE;
	[_exportController settingsFromDriverSettings:&exportSettings];
	ExportContextInfo *ei = calloc(sizeof(ExportContextInfo), 1);
	ei->tag = expType;
	ei->fileURL = CFBridgingRetain(theURL);
	
	[NSApp beginSheet:_exportWindow modalForWindow:[self windowForSheet] modalDelegate:self didEndSelector:@selector(exportSheetDidEnd:returnCode:contextInfo:) contextInfo:ei];
}

- (IBAction)exportMusicAs:(id)sender
{
	NSInteger tag = [sender tag];
	[_theDriver beginExport];
	
	NSString *musicName = self.musicName;
	
	switch (tag) {
		case -1:
			//AIFF
		{
			NSSavePanel *savePanel = [NSSavePanel savePanel];
			[savePanel setAllowedFileTypes:@[AVFileTypeAIFF]];
			[savePanel setCanCreateDirectories:YES];
			[savePanel setCanSelectHiddenExtension:YES];
			if (![musicName isEqualToString:@""]) {
				[savePanel setNameFieldStringValue:musicName];
			}
			[savePanel setPrompt:@"Export"];
			[savePanel setTitle:@"Export as AIFF audio"];
			[savePanel beginSheetModalForWindow:[self windowForSheet] completionHandler:^(NSInteger result) {
				if (result == NSFileHandlingPanelOKButton) {
					[self showExportSettingsWithExportType:-1 URL:[savePanel URL]];
				} else
					[_theDriver endExport];
			}];
			
		}
			break;
			
		case -2:
			//MP4
		{
			NSSavePanel *savePanel = [NSSavePanel savePanel];
			[savePanel setAllowedFileTypes:@[AVFileTypeAppleM4A]];
			[savePanel setCanCreateDirectories:YES];
			[savePanel setCanSelectHiddenExtension:YES];
			if (![musicName isEqualToString:@""]) {
				[savePanel setNameFieldStringValue:musicName];
			}
			[savePanel setPrompt:@"Export"];
			[savePanel setTitle:@"Export as MPEG-4 Audio"];
			[savePanel beginSheetModalForWindow:[self windowForSheet] completionHandler:^(NSInteger result) {
				if (result == NSFileHandlingPanelOKButton) {
					[self showExportSettingsWithExportType:-2 URL:[savePanel URL]];
				} else
					[_theDriver endExport];
			}];
		}
			break;
			
		case -3:
			//MADK
		{
			NSSavePanel *savePanel = [NSSavePanel savePanel];
			[savePanel setAllowedFileTypes:@[MADNativeUTI]];
			[savePanel setCanCreateDirectories:YES];
			[savePanel setCanSelectHiddenExtension:YES];
			if (![musicName isEqualToString:@""]) {
				[savePanel setNameFieldStringValue:musicName];
			}
			[savePanel setPrompt:@"Export"];
			[savePanel setTitle:@"Export as MADK"];
			[savePanel beginSheetModalForWindow:[self windowForSheet] completionHandler:^(NSInteger result) {
				if (result == NSFileHandlingPanelOKButton) {
					PPExportObject *expObj = [[PPExportObject alloc] initWithDestination:[savePanel URL] exportBlock:^OSErr(NSURL *theURL, NSString *__autoreleasing *errStr) {
						OSErr theErr = [_theMusic exportMusicToURL:theURL];
						[_theDriver endExport];
						return theErr;

					}];
					[[NSApp delegate] addExportObject:expObj];
				} else {
					[_theDriver endExport];
				}
			}];
		}
			break;
			
		default:
		{
			if (tag > [globalMadLib pluginCount] || tag < 0) {
				NSBeep();
				[_theDriver endExport];
#if 0
				if (isQuitting) {
					[NSApp replyToApplicationShouldTerminate:YES];
				}
#endif
				
				return;
			}
			NSSavePanel *savePanel = [NSSavePanel savePanel];
			PPLibraryObject *tmpObj = [globalMadLib pluginAtIndex:tag];
			[savePanel setAllowedFileTypes:tmpObj.UTItypes];
			[savePanel setCanCreateDirectories:YES];
			[savePanel setCanSelectHiddenExtension:YES];
			if (![musicName isEqualToString:@""])
				[savePanel setNameFieldStringValue:musicName];
			
			[savePanel setPrompt:@"Export"];
			[savePanel setTitle:[NSString stringWithFormat:@"Export as %@", tmpObj.menuName]];
			[savePanel beginSheetModalForWindow:[self windowForSheet] completionHandler:^(NSInteger result) {
				if (result == NSFileHandlingPanelOKButton) {
					PPExportObject *expObj = [[PPExportObject alloc] initWithDestination:[savePanel URL] exportBlock:^OSErr(NSURL *theURL, NSString *__autoreleasing *errStr) {
						OSErr theErr = [_theMusic exportMusicToURL:theURL format:tmpObj.plugType library:globalMadLib];
						[_theDriver endExport];
						return theErr;
					}];
					[[NSApp delegate] addExportObject:expObj];
				} else
					[_theDriver endExport];
			}];
		}
			break;
	}
}

- (IBAction)showBoxEditor:(id)sender
{
	
}

- (IBAction)showClassicEditor:(id)sender
{
	
}

- (IBAction)showDigitalEditor:(id)sender
{
	
}

- (IBAction)showWavePreview:(id)sender
{
	
}

- (IBAction)okayExportSettings:(id)sender {
	[NSApp endSheet:self.exportWindow returnCode:NSAlertDefaultReturn];
	[self.exportWindow close];
}

- (IBAction)cancelExportSettings:(id)sender {
	[NSApp endSheet:self.exportWindow returnCode:NSAlertAlternateReturn];
	[self.exportWindow close];
}

- (void)awakeFromNib
{
	[self.exportSettingsBox setContentView:self.exportController.view];
}

- (void)dealloc
{
	[[NSNotificationCenter defaultCenter] removeObserver:self];
}

#pragma mark PPSoundSettingsViewControllerDelegate methods

- (void)soundOutBitsDidChange:(short)bits
{
	exportSettings.outPutBits = bits;
}

- (void)soundOutRateDidChange:(unsigned int)rat
{
	exportSettings.outPutRate = rat;
}

- (void)soundOutReverbDidChangeActive:(BOOL)isAct
{
	exportSettings.Reverb = isAct;
}

- (void)soundOutOversamplingDidChangeActive:(BOOL)isAct
{
	if (!isAct) {
		exportSettings.oversampling = 1;
	}
}

- (void)soundOutStereoDelayDidChangeActive:(BOOL)isAct
{
	if (!isAct) {
		exportSettings.MicroDelaySize = 0;
	}
}

- (void)soundOutSurroundDidChangeActive:(BOOL)isAct
{
	exportSettings.surround = isAct;
}

- (void)soundOutReverbStrengthDidChange:(short)rev
{
	exportSettings.ReverbStrength = rev;
}

- (void)soundOutReverbSizeDidChange:(short)rev
{
	exportSettings.ReverbSize = rev;
}

- (void)soundOutOversamplingAmountDidChange:(short)ovs
{
	exportSettings.oversampling = ovs;
}

- (void)soundOutStereoDelayAmountDidChange:(short)std
{
	exportSettings.MicroDelaySize = std;
}

@end

File: PlayerPRO 6/Document classes/PPDocument.swift
//
//  PPDocument.swift
//  PPMacho
//
//  Created by C.W. Betts on 8/21/14.
//
//

import Cocoa
import PlayerPROCore
import PlayerPROKit
import AVFoundation
import AudioToolbox

@objc(PPDocument) class PPDocument: NSDocument {	
	var instrumentList: InstrumentPanelController! = nil
	var mainViewController: DocumentWindowController! = nil
	@objc dynamic let theDriver: PPDriver
	@objc dynamic private(set) var theMusic: PPMusicObject!
	
	@objc dynamic var musicName: String {
		get {
			return theMusic.title
		}
		set {
			theMusic.title = newValue
		}
	}
	
	@objc dynamic var musicInfo: String {
		get {
			return theMusic.information
		}
		set {
			theMusic.information = newValue
		}
	}
	
	// MARK: - NSDocument type info
	override class func isNativeType(_ aType: String) -> Bool {
		return aType == MADNativeUTI
	}
	
	override class var readableTypes: [String] {
		struct StaticStorage {
			static var readables: [String]?
		}
		
		if let readables = StaticStorage.readables {
			return readables
		}
		
		let importUTIsArray = globalMadLib.filter({ $0.canImport == true }).map({ $0.UTITypes })
		let importUTISet: Set<String> = {
			var toRet = Set<String>()
			for arr in importUTIsArray {
				toRet.formUnion(arr)
			}
			return toRet
		}()
		var toRet = Array(importUTISet)
		toRet.insert(MADNativeUTI, at: 0)
		toRet.append(MADGenericUTI)
		StaticStorage.readables = toRet
		return toRet
	}
	
	override class var writableTypes: [String] {
		struct StaticStorage {
			static var writables: [String]?
		}
		
		if let writables = StaticStorage.writables {
			return writables
		}
		
		var toRet = globalMadLib.filter({ $0.canExport == true }).map({ $0.UTITypes.first! })
		toRet.insert(MADNativeUTI, at: 0)
		StaticStorage.writables = toRet
		return toRet
	}
	
	// MARK: -
	
	override func makeWindowControllers() {
		let docWinCon = DocumentWindowController(windowNibName: NSNib.Name(rawValue: "PPDocument"))
		addWindowController(docWinCon)
		docWinCon.currentDocument = self
		instrumentList = InstrumentPanelController(windowNibName: NSNib.Name(rawValue: "InsPanel"))
		addWindowController(instrumentList)
		instrumentList.currentDocument = self
		mainViewController = docWinCon
	}

	private func resetPlayerPRODriver() {
		var theSett = MADDriverSettings.new()
		let defaults = UserDefaults.standard
		
		//TODO: Sanity Checking
		theSett.surround = defaults.bool(forKey: PPSurroundToggle)
		theSett.outPutRate = UInt32(defaults.integer(forKey: PPSoundOutRate))
		theSett.outPutBits = Int16(defaults.integer(forKey: PPSoundOutBits))
		if (defaults.bool(forKey: PPOversamplingToggle)) {
			theSett.oversampling = Int32(defaults.integer(forKey: PPOversamplingAmount))
		} else {
			theSett.oversampling = 1;
		}
		theSett.Reverb = defaults.bool(forKey: PPReverbToggle)
		theSett.ReverbSize = Int32(defaults.integer(forKey: PPReverbAmount))
		theSett.ReverbStrength = Int32(defaults.integer(forKey: PPReverbStrength))
		if (defaults.bool(forKey: PPStereoDelayToggle)) {
			theSett.MicroDelaySize = Int32(defaults.integer(forKey: PPStereoDelayAmount))
		} else {
			theSett.MicroDelaySize = 0;
		}
		
		theSett.driverMode = MADSoundOutput(rawValue: Int16(defaults.integer(forKey: PPSoundDriver))) ?? .CoreAudioDriver
		theSett.repeatMusic = false;
		
		do {
			try theDriver.changeDriverSettings(to: &theSett)
		} catch {
			Swift.print("Unable to change driver for \(self), error '\(error)'")
			//NSAlert(error: createErrorFromMADErrorType(returnerr)).beginSheetModalForWindow(self.windowForSheet, completionHandler: { (returnCode) -> Void in
			//currently, do nothing
			//})

		}
	}
	
	@objc private func soundPreferencesDidChange(_ notification: Notification) {
		resetPlayerPRODriver()
	}
	
	convenience init(music: PPMusicObject) {
		self.init()
		theMusic = music
	}
	
	override init() {
		var drivSettings = MADDriverSettings.new()
		let defaults = UserDefaults.standard
		
		//TODO: Sanity Checking
		drivSettings.surround = defaults.bool(forKey: PPSurroundToggle)
		drivSettings.outPutRate = UInt32(defaults.integer(forKey: PPSoundOutRate))
		drivSettings.outPutBits = Int16(defaults.integer(forKey: PPSoundOutBits))
		if defaults.bool(forKey: PPOversamplingToggle) {
			drivSettings.oversampling = Int32(defaults.integer(forKey: PPOversamplingAmount))
		} else {
			drivSettings.oversampling = 1
		}
		drivSettings.Reverb = defaults.bool(forKey: PPReverbToggle)
		drivSettings.ReverbSize = Int32(defaults.integer(forKey: PPReverbAmount))
		drivSettings.ReverbStrength = Int32(defaults.integer(forKey: PPReverbStrength))
		if (defaults.bool(forKey: PPStereoDelayToggle)) {
			drivSettings.MicroDelaySize = Int32(defaults.integer(forKey: PPStereoDelayAmount))
		} else {
			drivSettings.MicroDelaySize = 0;
		}
		
		drivSettings.driverMode = MADSoundOutput(rawValue: Int16(defaults.integer(forKey: PPSoundDriver))) ?? .CoreAudioDriver
		drivSettings.repeatMusic = false;
		
		theDriver = try! PPDriver(library: globalMadLib, settings: &drivSettings)
		super.init()
		
		let defaultCenter = NotificationCenter.default
		defaultCenter.addObserver(self, selector: #selector(PPDocument.soundPreferencesDidChange(_:)), name: .PPSoundPreferencesDidChange, object: nil)
	}
	
    override func windowControllerDidLoadNib(_ aController: NSWindowController) {
        super.windowControllerDidLoadNib(aController)
        // Add any code here that needs to be executed once the windowController has loaded the document's window.
    }

	override func write(to url: URL, ofType typeName: String) throws {
		if typeName != MADNativeUTI {
			guard let type = globalMadLib.typeFromUTI(typeName) else {
				throw NSError(domain: NSOSStatusErrorDomain, code: paramErr, userInfo: nil)
			}
			try theMusic.exportMusic(to: url, format: type, library: globalMadLib)
		} else {
			do {
				try theMusic.saveMusic(to: url, compress: UserDefaults.standard.bool(forKey: PPMMadCompression))
			} catch {
				if let error = error as? MADErr {
					throw error.convertToCocoaType()
				} else {
					throw error
				}
			}
		}
	}
	
	//override func write(to url: URL, ofType typeName: String, for saveOperation: NSSaveOperationType, originalContentsURL absoluteOriginalContentsURL: URL?) throws {
	//	try super.write(to: url, ofType: typeName, for: saveOperation, originalContentsURL: absoluteOriginalContentsURL)
	//}
	
	override func read(from url: URL, ofType typeName: String) throws {
		func getType() throws -> String {
			if let type = globalMadLib.typeFromUTI(typeName) {
				return type
			}
			
			return try globalMadLib.identifyFile(at: url)
		}
		
		if typeName == MADNativeUTI {
			theMusic = try PPMusicObject(url: url, driver: theDriver)
		} else {
			let theType = try getType()
			theMusic = try PPMusicObject(url: url, stringType: theType, driver: theDriver)
			fileURL = nil
		}
	}
	
	override var autosavingFileType: String? {
		return MADNativeUTI
	}
	
	override class var autosavesInPlace: Bool {
		return true
	}
	/*
    override class func autosavesInPlace() -> Bool {
        return true
    }*/

	func importMusicObject(_ theObj: PPMusicObject) {
		if theMusic == nil {
			theMusic = theObj
		}
	}
		
	deinit {
		NotificationCenter.default.removeObserver(self)
	}
}
